docs/src/dev.md

```diff
@@ -0,0 +1,55 @@
+# Developer Notes
+
+Although, the mechanism is simple, the fragile type inference of Julia compiler bring this package to a lot more complicated
+performance optimization issue.
+
+## Fixing Type Inference Failures
+The best way to find type inference failure and fix it is using [Cthulhu.jl](https://github.com/JuliaDebug/Cthulhu.jl). Once you find a type inference failure, you will need to put them into the Cassette `overdub` pass, e.g
+
+If we find type inference fails for `Base.iterate`, the following function should be defined
+
+```julia
+@inline Cassette.overdub(ctx::RecordingCtx, ::typeof(Base.iterate), it, st) = iterate(it, st)
+@inline Cassette.overdub(ctx::ReplayCtx, f::typeof(Base.iterate), it, st) = iterate(it, st)
+```
+
+where [`RecordingCtx`](@ref) is the Cassette context for the first tracing execution, and the [`ReplayCtx`](@ref) is the actual
+runtime Cassette context. You could also read [`inference_fixes.jl`]() for more details.
+
+### Working with Zygote
+To work with Zygote, since Zygote will use [IRTools](https://github.com/MikeInnes/IRTools.jl) to implement contextual dispatch,
+the type inference failure may be different. However, one should always remember to pass the most type inference failure function to the following functions as well
+
+```julia
+@inline Cassette.overdub(ctx::RecordingCtx, ::typeof(Zygote._pullback), cx::Zygote.AContext, ::typeof(Base.iterate), it, st) = Zygote._pullback(cx, Base.iterate, it, st)
+@inline Cassette.overdub(ctx::ReplayCtx, ::typeof(Zygote._pullback), cx::Zygote.AContext, ::typeof(Base.iterate), it, st) = Zygote._pullback(cx, Base.iterate, it, st)
+```
+
+## Performance Tips
+
+### Transform operations like `pop!`, `push!`
+
+In many cases, one can write `pop!`, or `push!` as a locally preallocated but uninitialized version, e.g
+
+if we the total length of the list is known, we could preallocate the memory in advance, instead of writing
+
+```julia
+list = []
+for k in 1:L
+    x = #= some computation =#
+    push!(list, x)
+end
+```
+
+we could write
+
+```julia
+list = Vector{Any}(undef, L)
+for k in 1:L
+    x = #= some computation =#
+    list[k] = x
+end
+```
+
+then we won't be resizing the `Vector` during runtime, which make it much easier to preallocate everything in the record created
+by `AutoPreallocation`.
```

Comment:
Resizing vectors is also permitted by `AutoPrecallocation` though and that should be mentioned here.
We have a function that takes records and fixs them  via `resize`

Reply:
you mean `reinitialize!`? but I think once this is used the `push!` will still reallocate using `resize!`?

Reply:
Does `resize!` ing down decrease memory used by an array?
my reading of the C is that it doesn’t since i don’t ssee any references to updating `maxsize` only `nrows`
https://github.com/JuliaLang/julia/blob/7426625b5c07b0d93110293246089a259a0a677d/src/array.c#L1056

So it *shouldn't allocate to `push!` again, because effectively have already done `sizehint`

Reply:
I see. sorry to reply so late. Would you mind to revise this part directly in this PR? I'm not sure how to mention it here actually.